File: argo/pca.R
source("read.R")

library(ggplot2)
library(ggrastr)
library(dr)

m = dim(temp)[1] # Number of pressure points
n = dim(temp)[2] # Number of profiles

# The mean profile
tempmean = apply(temp, 1, mean)
psalmean = apply(psal, 1, mean)

# Center the profiles
tempc = temp - outer(tempmean, array(1, n))
psalc = psal - outer(psalmean, array(1, n))

# Get the principal components
cc = cov(t(tempc))
ee = eigen(cc, symmetric=T)
eigval = ee$values
eigvec = ee$vectors

# Flip each PC loading vector if it is mostly negative.
for (j in 1:dim(eigvec)[2]) {
    if (sum(eigvec[,j] < 0) > sum(eigvec[,j] > 0)) {
        eigvec[,j] = -eigvec[,j]
    }
}

# Scores for the dominant PC's
scores = t(tempc) %*% eigvec[,1:10]

pdf("pca_r.pdf")

# Plot the mean profile
da = data.frame(pressure=pressure, tempmean=tempmean)
plt = ggplot(aes(x=pressure, y=tempmean), data=da) + geom_line()
print(plt)

# Make several plots depicting the loading patterns.
for (j in 1:5) {
    # Plot the loadings
    da = data.frame(pressure=pressure, loading=eigvec[,j])
    plt = ggplot(aes(x=pressure, y=loading), data=da) + geom_line()
    plt = plt + labs(x="Pressure", y="Temperature loading")
    plt = plt + ggtitle(sprintf("PC %d", j))
    print(plt)

    # Plot the mean profile +/- multiples of the loadings
    da = data.frame()
    s = sd(scores[,j])
    for (f in c(-2, -1, 0, 1, 2)) {
        dx = data.frame(pressure=pressure, profile=tempmean+f*s*eigvec[,j], f=f)
        da = rbind(da, dx)
    }
    da$f = as.factor(da$f)
    plt = ggplot(aes(x=pressure, y=profile, color=f, group=f), data=da) + geom_line()
    plt = plt + labs(x="Pressure", y="Temperature")
    plt = plt + ggtitle(sprintf("PC %d", j))
    print(plt)
}

# Plot the j^th PC score against the k^th feature.
plot_pcscores = function(j, k) {

    xx = seq(min(Y[,k]), max(Y[,k]), length=100)
    m = lowess(Y[,k], scores[,j])
    mf = approxfun(m$x, m$y)
    f = abs(m$y - scores[,j])
    r = lowess(Y[,k], f)
    rf = approxfun(r$x, r$y)
    da = data.frame(x=xx, y=mf(xx), r=rf(xx))

    f = 2
    da$y1 = da$y - f*da$r
    da$y2 = da$y + f*da$r
    plt = ggplot(aes(x=x, y=y), data=da)
    plt = plt + labs(x=fn[k], y=sprintf("PC %d score", j))
    plt = plt + geom_ribbon(aes(x=x, ymin=y1, ymax=y2), fill="grey70")
    plt = plt + geom_line()
    print(plt)
}

fn = c("Latitude", "Longitude", "Day")
Y = cbind(lat, lon, day)

for (j in 1:3) {
    for (k in 1:3) {
        plot_pcscores(j, k)
    }
}

# Flip signs in two sets of CCA loadings to make them easier to understand in a
# plot.
flip = function(xcoef, ycoef) {
    for (j in 1:dim(xcoef)[2]) {
        if (mean(xcoef[,j] > 0) + mean(ycoef[,j] > 0) < 1) {
            xcoef[,j] = -xcoef[,j]
            ycoef[,j] = -ycoef[,j]
        }
    }
    return(list(xcoef=xcoef, ycoef=ycoef))
}

# Use a combined PCA/CCA approach to relate temperature and salinity.
# Produce plots for a sequence of dimensions for the PCA dimension
# reduction.
plot_cca = function() {
    X = t(tempc)
    Y = t(psalc)
    svx = svd(X)
    svy = svd(Y)

    for (q in c(1, 2, 5, 10)) {
        cc = cancor(svx$u[,1:q], svy$u[,1:q])
        rr = cc$cancorr

        # Passing nrow and ncol is needed to avoid
        # different behavior when q=1.
        ddx = diag(svx$d[1:q], q, q)
        ddy = diag(svy$d[1:q], q, q)

        xcoef = svx$v[,1:q] %*% solve(ddx, cc$xcoef)
        ycoef = svy$v[,1:q] %*% solve(ddy, cc$ycoef)
        ll = flip(xcoef, ycoef)
        xcoef = ll$xcoef
        ycoef = ll$ycoef
        da = data.frame(temp_weights=xcoef[,1], psal_weights=ycoef[,1], pressure=pressure)
        plt = ggplot(aes(x=pressure, y=temp_weights), data=da) + geom_line()
        plt = plt + labs(x="Pressure", y="Temperature coefficient")
        plt = plt + ggtitle(sprintf("CCA with q=%d components, r=%.2f", q, rr[1]))
        if (min(ycoef[,1]) > 0) {
            plt = plt + ylim(0, NA)
        }
        print(plt)
        plt = ggplot(aes(x=pressure, y=psal_weights), data=da) + geom_line()
        plt = plt + labs(x="Pressure", y="Salinity coefficient")
        plt = plt + ggtitle(sprintf("CCA with q=%d components, r=%.2f", q, rr[1]))
        if (min(ycoef[,1]) > 0) {
            plt = plt + ylim(0, NA)
        }
        print(plt)
    }
}

plot_cca()

# SIR is a dimension reduction regression (DR) method.  Here we will use
# it doidentify factors within the temperature data that predict latitude.
plot_sir = function() {

    X = t(tempc)

    # Due to high-dimensionality, we project the temperature data to
    # a limited number of PC's before using SIR to estimate the coefficients.
    q = 5
    svx = svd(X)
    dd = dr.compute(svx$u[,1:q], lat, array(1, length(lat)))
    ddx = diag(svx$d[1:q], q, q)
    b = svx$v[,1:q] %*% solve(ddx, dd$evectors)

    # Plot the loadings
    da = data.frame(pressure=c(pressure, pressure, pressure))
    da$dir = c(b[,1], b[,2], b[,3])
    oo = array(1, length(pressure))
    da$group = c(1*oo, 2*oo, 3*oo)
    da$group = as.factor(da$group)
    plt = ggplot(aes(x=pressure, y=dir, by=group, color=group), data=da) + geom_line()
    plt = plt + labs(x="Pressure", y="Temperature loading")
    plt = plt + ggtitle(sprintf("SIR with q=%d PC components", q))
    print(plt)

    # Plot the scores against latitude, longitude, and day.
    scores = X %*% b[,1:3]
    for (v in c("lat", "lon", "day")) {
        for (j in 1:3) {
            da = data.frame(lat=lat, lon=lon, day=day, s=scores[,j])
            plt = ggplot(aes(x=!!sym(v), y=s), data=da) + geom_point()
            plt = plt + labs(x=v, y=sprintf("SIR component %d", j))
            rasterize(plt, layers="Point")
            print(plt)
        }
    }
}

da = plot_sir()

dev.off()

File: argo/pca.py
import numpy as np
import pandas as pd
from matplotlib.backends.backend_pdf import PdfPages
import matplotlib.pyplot as plt
import os
from scipy.interpolate import interp1d
from statsmodels.nonparametric.smoothers_lowess import lowess
import statsmodels.api as sm
from statsmodels.regression.dimred import SIR
from read import *

pdf = PdfPages("pca_py.pdf")

# Convert the date to the number of days since the first date
# in the dataset.
date = pd.to_datetime(date)
ddate = date - date.min()
day = [x.days for x in ddate]

# Get the circular mean of all longitude values.
lon_sin = np.sin(2*np.pi*lon/360)
lon_cos = np.cos(2*np.pi*lon/360)
cmean_lon = np.arctan2(lon_sin.mean(), lon_cos.mean()) * 360 / (2*np.pi)

# Convert latitude so that the Pacific ocean doesn't wrap
# around the origin.
lon = (lon + 60) % 360

# Create a matrix of observed variables that describe
# the location and time at which each profile was
# obtained.
n = len(lat)
Y = np.zeros((n, 3))
Y[:, 0] = lat
Y[:, 1] = lon
Y[:, 2] = day

plt.clf()
plt.grid(True)
plt.plot(lon, lat, "o", rasterized=True, alpha=0.2)
plt.xlabel("Longitude (translated)", size=15)
plt.ylabel("Latitude", size=15)
plt.title("Float positions")
pdf.savefig()

# Get the principal components.
def get_pcs(x):
    xc = x.copy()
    xm = x.mean(1)
    for j in range(x.shape[0]):
        xc[j, :] = x[j, :] - xm[j]
    cc = np.cov(xc)
    pcw, pcv = np.linalg.eigh(cc)

    # Reorder the PC's so that the dominant factors
    # are first.
    ii = np.argsort(pcw)[::-1]
    pcw = pcw[ii]
    pcv = pcv[:, ii]

    # For interpretability flip the PC's that are
    # mostly negative.
    for j in range(pcv.shape[1]):
        if (pcv[:, j] < 0).sum() > (pcv[:, j] >= 0).sum():
            pcv[:, j] *= -1

    # Get the PC scores for temperature
    scores = np.dot(xc.T, pcv[:, 0:5])

    return xm, pcw, pcv, scores

# Plot the j^th PC score against all features.
def pcplot(j, mean, pcv, scores, label):

    # Plot the mean profile
    plt.clf()
    plt.grid(True)
    plt.plot(pressure, mean)
    plt.gca().set_xlabel("Pressure", size=15)
    plt.gca().set_ylabel("Mean %s" % label, size=15)
    pdf.savefig()

    # Plot the PC loadings
    fn = ["Latitude", "Longitude", "Day"]
    plt.clf()
    plt.grid(True)
    plt.plot(pressure, pcv[:, j])
    plt.gca().set_xlabel("Pressure", size=15)
    plt.gca().set_ylabel("%s PC %d loading" % (label.title(), j + 1), size=15)
    if pcv[:, j].min() > 0:
        plt.gca().set_ylim(ymin=0)
    pdf.savefig()

    # Plot the mean +/- 1 SD of the loading pattern
    plt.clf()
    plt.title(label.title())
    plt.grid(True)
    s = scores[:, j].std()
    for f in [-1, 0, 1]:
        plt.plot(pressure, mean + f*s*pcv[:, j], color={-1: "blue", 0: "black", 1: "red"}[f])
    plt.gca().set_xlabel("Pressure", size=15)
    plt.gca().set_ylabel("Mean %s +/- PC %d loading" % (label, j + 1), size=15)
    pdf.savefig()

    # Plot the conditional mean PC score against an observed variable,
    # showing the conditional mean plus/minus one absolute deviation.
    for k in range(3):
        xx = np.linspace(Y[:, k].min(), Y[:, k].max(), 100)
        m = lowess(scores[:, j], Y[:, k], delta=0.01*np.ptp(Y[:, k]))
        resid = scores[:, j] - m[:, 1]
        r = lowess(np.abs(resid), Y[:, k], delta=0.01*np.ptp(Y[:, k]))
        yy = interp1d(m[:, 0], m[:, 1])(xx)
        yr = interp1d(r[:, 0], r[:, 1])(xx)
        f = 2
        ymx = (yy + f*yr).max()
        ymn = (yy - f*yr).min()
        plt.clf()
        plt.grid(True)
        plt.plot(xx, yy, "-", color="red")
        plt.plot(xx, yy-f*yr, "-", color="grey")
        plt.plot(xx, yy+f*yr, "-", color="grey")
        plt.gca().set_ylim([ymn, ymx])
        plt.gca().set_xlabel(fn[k], size=15)
        plt.gca().set_ylabel("%s PC %d score" % (label.title(), j + 1), size=15)
        pdf.savefig()

tempmean, tempw, tempv, tempscores = get_pcs(temp)
psalmean, psalw, psalv, psalscores = get_pcs(psal)

for j in range(3):
    pcplot(j, tempmean, tempv, tempscores, "temperature")

for j in range(3):
    pcplot(j, psalmean, psalv, psalscores, "salinity")

# CCA that agrees with R.
def my_cca(X, Y):
    n = X.shape[0]
    X = X - X.mean(0)
    Y = Y - Y.mean(0)
    Sx = np.dot(X.T, X) / n
    Sy = np.dot(Y.T, Y) / n
    Sxy = np.dot(X.T, Y) / n
    Rx = np.linalg.cholesky(Sx)
    Ry = np.linalg.cholesky(Sy)
    M = np.linalg.solve(Rx, Sxy)
    M = np.linalg.solve(Ry, M.T).T
    u, s, vt = np.linalg.svd(M)
    v = vt.T
    u = np.linalg.solve(Rx.T, u)
    v = np.linalg.solve(Ry.T, v)
    return u, v, s


# Standard CCA, due to the high dimensionality and consequent
# overfitting, the results make little sense.
X = temp.T.copy()
X -= X.mean(0)
Y = psal.T.copy()
Y -= Y.mean(0)
xc, yc, r = my_cca(X, Y)

# Flip the CCA components as needed for interpretability
def flip(xc, yc):
    for j in range(xc.shape[1]):
        if (xc[:, j] > 0).mean() + (yc[:, j] > 0).mean() < 1:
            xc[:, j] *= -1
            yc[:, j] *= -1
    return xc, yc

# The columns of ux and uy are the PC's of the temperature and salinity data.
ux,sx,vtx = np.linalg.svd(X, 0)
uy,sy,vty = np.linalg.svd(Y, 0)

# Reduce the temperature and salinity data to PCs, then do CCA on
# the projected data and finally map the loadings back to the original
# coordinates (this is very similar to PCR but applied to CCA not to
# linear regression).
for q in [1, 2, 5, 10, 20, 50]:

    # Do CCA after projecting the profiles to the top q PC's.
    xc, yc, r = my_cca(ux[:, 0:q], uy[:, 0:q])

    # Map the loadings back to the original coordinates
    xc1 = np.dot(vtx.T[:, 0:q], np.linalg.solve(np.diag(sx[0:q]), xc))
    yc1 = np.dot(vty.T[:, 0:q], np.linalg.solve(np.diag(sy[0:q]), yc))
    xc1, yc1 = flip(xc1, yc1)
    print(np.std(xc1[:, 0]))
    print(np.std(yc1[:, 0]))

    # Plot the temperature loadings
    plt.clf()
    plt.axes([0.15, 0.1, 0.8, 0.8])
    plt.grid(True)
    plt.title("CCA/PCA using %d principal components, r=%.2f" % (q, r[0]))
    plt.plot(pressure, xc1[:, 0])
    if xc1[:, 0].min() > 0:
        plt.ylim(ymin=0)
    plt.xlabel("Pressure", size=15)
    plt.ylabel("Temperature loading", size=15)
    pdf.savefig()

    # Plot the salinity loadings
    plt.clf()
    plt.axes([0.15, 0.1, 0.8, 0.8])
    plt.title("CCA/PCA using %d principal components, r=%.2f" % (q, r[0]))
    plt.grid(True)
    plt.plot(pressure, yc1[:, 0])
    if yc1[:, 0].min() >= 0:
        plt.ylim(ymin=0)
    plt.xlabel("Pressure", size=15)
    plt.ylabel("Salinity loading", size=15)
    pdf.savefig()
    print(np.corrcoef(np.dot(X, xc1[:, 0]), np.dot(Y, yc1[:, 0]))[0,1])

# Use Sliced Inverse Regression to predict latitude from the first
# q principal components of the temperature data.
q = 5
m = SIR(lat, ux[:, 0:q])
r = m.fit()
cf = np.dot(vtx.T[:, 0:q], np.linalg.solve(np.diag(sx[0:q]), r.params))

# Plot the SIR loadings.
plt.clf()
plt.grid(True)
for j in range(3):
    plt.plot(pressure, cf[:, j], "-", label="%d" % (j + 1))
ha, lb = plt.gca().get_legend_handles_labels()
leg = plt.figlegend(ha, lb, loc="center right")
leg.draw_frame(False)
plt.xlabel("Pressure", size=15)
plt.ylabel("SIR loading", size=15)
pdf.savefig()

# Plot the SIR scores against latitude.
scores = np.dot(X, cf)
for j in range(3):
    plt.clf()
    plt.grid(True)
    plt.plot(lat, scores[:, j], "o", color="grey", alpha=0.3, rasterized=True)

    # Use lowess to estimate the conditional mean of the scores given latitude.
    # Lowess is slow and doesn't need all the data to give an accurate estimate.
    ii = np.random.choice(np.arange(scores.shape[0]), 2000, replace=False)
    m = lowess(scores[ii, j], lat[ii], frac=0.2)
    plt.plot(m[:, 0], m[:, 1], "-", color="orange")

    plt.xlabel("Latitude", size=15)
    plt.ylabel("Component %d SIR score" % (j + 1), size=15)
    pdf.savefig()

pdf.close()
